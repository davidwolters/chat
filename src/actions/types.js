

export const USER_EXISTS = 	'user-exists';
export const USER_LOGIN = 'login';
export const USER_REGISTER = 'register';

export const FRIENDS_GET = 'get-friends';
export const FRIENDS_SEND_REQUEST = 'send-friend-request';
export const FRIENDS_ACCEPT_REQUEST = 'accept-friend-request';

export const MESSAGES_GET = 'get-conversation';
export const MESSAGES_SEND = 'send-message';

export const NAVIGATE_TO = 'navigate-to';
export const ENTER_CHAT = 'enter-chat';