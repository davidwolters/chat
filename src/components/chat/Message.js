import styled from 'styled-components'
import ThemeContext from '../../global/Theme'
import React, { useContext } from 'react'

const MessageBubble = styled.div`
	border-radius: 50px;
	display: inline-block;
	max-width: 50%;
	
	text-align: right;
	
	background-color: ${props => props.bg};
	color: white;
	padding: 5px 10px;
	font-family: ${props => props.theme.font.body.family};
	font-weight: ${props => props.theme.font.body.weight};
`

const MessageContainer = styled.div`
	display: block;
	text-align: ${props => props.textAlign};
	margin-bottom: 5px;
`

const Message = props => {

	const theme = useContext(ThemeContext)

	// If the message is outgoing, we use the accent color, if not, we use the
	// disabled color, and adjust textAlign accordingly.
	const background = props.outgoing === true
		? theme.color.accent
		: theme.color.disabled
	const textAlign = props.outgoing === true ? 'right' : 'left'

	return (
		<div>
			<MessageContainer textAlign={textAlign}>
				<MessageBubble bg={background}>{props.message}</MessageBubble>
			</MessageContainer>
		</div>
	)
}

export default Message