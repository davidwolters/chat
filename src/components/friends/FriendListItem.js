import styled from 'styled-components'
import React, { useContext } from 'react'
import ThemeContext from '../../global/Theme'

export const WAITING_FOR_APPROVAL = 0
export const WAITING_FOR_ACCEPT = 1
export const CONFIRMED = 2

// Container element.
const FriendContainer = styled.div`
	background-color: ${props => props.theme.color.windowBg};
	padding: 10px 30px;
	border-radius: 20px;
	margin-bottom: 10px;
	position: relative;
	margin-top: 0;	
	color: ${props => props.theme.color.text};
	&:hover {
		color: ${props => props.theme.color.accent};
		cursor: pointer;
	}
`

// Title.
export const FriendTitle = styled.h2`
	font-size: ${props => props.theme.font.size.title};
	font-family: ${props => props.theme.font.body.family};
	display: inline-block;
	margin: 0;
	padding: 0;
`

// The actual status button.
const FriendStatusButton = styled.div`
	float: right;
	border-radius: 50px;
	font-size: ${props => props.theme.font.size.text};
	font-weight: ${props => props.theme.font.title.weight};
	font-family: ${props => props.theme.font.body.family};
	padding: 7px;
	background-color: ${props => props.bg};
	color: ${props => props.color};
	border: none;
	position: absolute;
	right: 0%;
	top: 50%;
	transform: translate(-50%, -50%);
	&:hover {
		background-color: ${props => props.activeColor};
	}
`

const FriendListItem = props => {

	// Use the theme for styling depending on status.
	const theme = useContext(ThemeContext)

	let statusButton = ''

	// If we have sent the request, and are waiting on the other to accept,
	// display gray waiting button
	if (props.status === WAITING_FOR_APPROVAL) {
		statusButton = <FriendStatusButton bg={theme.color.disabled}
										   color={theme.color.text}
										   activeColor={theme.color.disabled}>Waiting...</FriendStatusButton>
	}
	else if (props.status === WAITING_FOR_ACCEPT) {
		// Otherwise, if we have recieved the request, display primary accept
		// button with onclick event
		statusButton = <FriendStatusButton bg={theme.color.primary}
										   color={'white'} onClick={() => {
			props.onAccept(props.friendID)
		}}
										   activeColor={theme.color.primaryDark}>Accept</FriendStatusButton>
	}
	else if (props.status === CONFIRMED) {
		// If we are already friends, display chat button with chat onclick
		// event
		statusButton = <FriendStatusButton bg={theme.color.accent} color='white'
										   onClick={() => props.enterChat(
											   props.friendID)}
										   activeColor={theme.color.accentDark}>Chat</FriendStatusButton>
	}
	return (
		<FriendContainer>
			<FriendTitle>{props.name}</FriendTitle>
			{statusButton}
		</FriendContainer>
	)
}

export default FriendListItem