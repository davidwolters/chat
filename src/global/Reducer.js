import {
	ENTER_CHAT,
	FRIENDS_GET, MESSAGES_GET,
	NAVIGATE_TO,
	USER_EXISTS,
	USER_LOGIN,
	USER_REGISTER,
} from '../actions/types'


// Reducer to update global state.
export const Reducer = (state, action) => {

	switch (action.type) {
		case USER_EXISTS:
			return userExists(state, action)
		case USER_REGISTER:
			return userRegistered(state, action)
		case USER_LOGIN:
			return userLogin(state, action);
		case NAVIGATE_TO:
			return navigateTo(state, action);
		case FRIENDS_GET:
			return loadFriends(state, action);
		case ENTER_CHAT:
			return enterChat(state, action);
		case MESSAGES_GET:
			return loadMessages(state, action);
		default:
			return state;
	}
}

// User exists action handler.
const userExists = (state, action) => {
	const currentPage = action.payload.user_exists ? '/login' : '/register'

	return {
		...state,
		auth: {
			...state.auth,
			exists: action.payload.user_exists,
			username: action.payload.username,
		},
		nav: {
			...state.nav,
			currentPage,
		},
	}
}

// User registered action handler.
const userRegistered = (state, action) => {
	const currentPage = '/friends'
	return {
		...state,
		auth: {
			...state.auth,
			ID: action.payload.ID,
			password: action.payload.password
		},
		nav: {
			...state.nav,
			currentPage
		}
	}
}

// User logged in successfully action handler.
const userLogin = (state, action ) => {
	const currentPage = '/friends'
	return {
		...state,
		auth: {
			...state.auth,
			ID: action.payload.ID,
			password: action.payload.password,
			meta: action.payload.meta
		},
		friends: {
			...state.friends,
			friends: action.payload.friends.friends,
			incoming: action.payload.friends.incoming,
			outgoing: action.payload.friends.outgoing
		},
		nav: {
			...state.nav,
			currentPage
		}
	}
}

// Navigation action handler.
const navigateTo = (state, action) => {
	return {
		...state,
		nav: {
			...state.nav,
			currentPage: action.payload
		}
	}
}

// Load friends action handler.
const loadFriends = (state, action) => {
	return {
		...state,
		friends: {
			...state.friends,
			friends: action.payload.friends,
			incoming: action.payload.incoming,
			outgoing: action.payload.outgoing
		},
	}
}

// Enter chat action handler.
const enterChat = (state, action) => {
	const currentPage = '/chat';
	return {
		...state,
		nav: {
			currentPage,
			currentChat: action.payload
		}
	};
}

// Load messages action handler.
const loadMessages = (state, action) => {
	return {
		...state,
		messages: {
			...state.messages,
			[action.payload.ID]: action.payload.messages
		}
	}
}