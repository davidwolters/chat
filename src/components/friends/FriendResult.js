import styled from 'styled-components'
import React, { useContext } from 'react'
import { FriendTitle } from './FriendListItem'
import { FormButton } from '../form/FormElements'
import ThemeContext from '../../global/Theme'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck, faSpinner } from '@fortawesome/free-solid-svg-icons'

const FriendBubble = styled.div`
	background-color: white;
	padding: 10px 20px;
	border-radius: 10px;
	text-align: center;
	min-width: 30%;
	display: inline-block;
	
	
`

const FriendContainer = styled.div`
	width: 100%;
	height: 100%;
	display: block;
	position: relative;
	text-align: center;
	margin-top: 10px;
`

// FriendResult, showed in AddFriend.js
const FriendResult = props => {

	const theme = useContext(ThemeContext)

	// The button that lets us add the friend.
	let button = <FormButton color={theme.color.accent}
							 activeColor={theme.color.accentDark}
							 onClick={props.addFriend}>
		Add Friend</FormButton>

	// If we have clicked it, show that it is loading.
	if (props.status === 'loading') {
		button = (
			<FormButton color={theme.color.accent}
						activeColor={theme.color.accent}>
				<FontAwesomeIcon icon={faSpinner} pulse/>
			</FormButton>
		)
	}
	else if (props.status === 'added') {
		// If we have now added the friend, show confirmation.
		button = (
			<FormButton color={theme.color.secondary}
						activeColor={theme.color.secondary}>
				<FontAwesomeIcon icon={faCheck}/>
			</FormButton>
		)
	}
	else if (props.status === 'error') {
		// If something went wrong, show that.
		button = (
			<FormButton color={theme.color.error}
						activeColor={theme.color.error}>
				Error
			</FormButton>
		)
	}
	return (
		<div>
			<FriendContainer>
				<FriendBubble>
					<FriendTitle>{props.name}</FriendTitle>
					{button}
				</FriendBubble>
			</FriendContainer>
		</div>
	)
}

export default FriendResult