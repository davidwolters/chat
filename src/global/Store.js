import React, {useReducer, createContext } from 'react';
import {Reducer} from './Reducer';

// Default state.
const initialState = {
	auth: {
		ID: null,
		username: null,
		password: null,
		exists: false
	},
	nav: {
		currentPage: '/enter-username',
		currentChat: null
	},
	friends: {
		incoming: [],
		outgoing: [],
		friends: []
	},
	messages: null
}

// Global state context store.
const Store = ({children}) => {
	// Create the state, and render the children with the context provider.
	const [state, dispatch] = useReducer(Reducer, initialState);
	return (
		<Context.Provider value={[state, dispatch]}>
			{children}
		</Context.Provider>
	)
}

// The context.
export const Context = createContext(initialState);
export default Store;