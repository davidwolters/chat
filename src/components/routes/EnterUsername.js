import {
	FormInput,
	FormTitle,
	FormWindow,
	FormButton, FormError, FormSubtitle,
} from '../form/FormElements'
import { useRef, useState, useEffect, useContext } from 'react'
import Background from '../app/Background'
import { sendRequest } from '../../api/apiClient'
import { USER_EXISTS } from '../../actions/types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faSpinner } from '@fortawesome/free-solid-svg-icons'

import { Context } from '../../global/Store'


// Screen to enter username.
const EnterUsername = ({ onSubmit }) => {

	// The username
	const [username, setUsername] = useState(null)

	// If we are loading.
	const [loading, setLoading] = useState(false)

	// Input ref for getting value.
	const inputRef = useRef(null)

	// We only need the dispatch, since we don't use the state, we only update it.
	const [, dispatch] = useContext(Context)

	// Error.
	const [error, setError] = useState('')

	// When we submit the form, check if we exist & if so if we do, send us to login, if we don't send us to register.
	useEffect(() => {
		const usernameReturned = data => {

			setLoading(false)
			if (data.status === 'success') {
				dispatch({
					type: USER_EXISTS,
					payload: {
						...data.data,
						username,
					},
				})
			}
			else {
				setError('Something went wrong: ' + data.data)
			}
		}
		if (username !== null) {
			setLoading(true)
			sendRequest(USER_EXISTS, { username }, usernameReturned)
		}
	}, [username, dispatch])

	// Handle submit, if we an input value, set Username to it.
	const handleSubmit = e => {
		e.preventDefault()
		if (inputRef.current.value !== null && inputRef.current.value !== '') {
			setUsername(inputRef.current.value);
		}
	}

	// Content of the form.
	let formContent = (
		<>
			<FormTitle>
				Enter username
			</FormTitle>
			<FormSubtitle>Or think of a new one!</FormSubtitle>
			<form onSubmit={handleSubmit}>
				<FormInput autocomplete="username" placeholder="Enter username" ref={inputRef}/>
				<FormButton>
					Go
				</FormButton>
			</form>
		</>
	)

	// If we have already submitted, show loading icon.
	if (loading) {
		formContent = (
			<>
				<FontAwesomeIcon icon={faSpinner} size={'lg'} pulse/>
			</>
		)
	}

	return (
		<Background>
			<FormWindow>
				{formContent}
				<FormError>{error}</FormError>
			</FormWindow>
		</Background>
	)
}
export default EnterUsername

