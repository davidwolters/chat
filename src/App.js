import React, { useContext } from 'react'
import {
	BrowserRouter as Router,
	Route,
} from 'react-router-dom'
import AuthWrapper from './components/auth/AuthWrapper'
import EnterUsername from './components/routes/EnterUsername'
import Login from './components/routes/Login'
import Register from './components/routes/Register'
import Store from './global/Store'
import Friends from './components/routes/Friends'
import ThemeContext  from './global/Theme'
import { ThemeProvider } from 'styled-components'
import Chat from './components/routes/Chat'
import AddFriend from './components/routes/AddFriend'

// Main app.
const App = () => {

	// Get the theme for the themeprovider.
	const theme = useContext(ThemeContext);

	return (
		<Store>
			<Router>
				<AuthWrapper>
					<ThemeContext.Provider value={theme}>
						<ThemeProvider theme={theme}>
							<Route path="/enter-username">
								<EnterUsername/>
							</Route>
							<Route path="/login">
								<Login/>
							</Route>
							<Route path="/register">
								<Register/>
							</Route>
							<Route path="/friends">
								<Friends/>
							</Route>
							<Route path="/chat">
								<Chat />
							</Route>
							<Route path="/add-friend">
								<AddFriend />
							</Route>
						</ThemeProvider>
					</ThemeContext.Provider>
				</AuthWrapper>

			</Router>
		</Store>
	)
}
export default App
