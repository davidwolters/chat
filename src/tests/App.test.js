import React, { useContext } from 'react'
import renderer from 'react-test-renderer'
import App from '../App'
import { act } from '@testing-library/react'
import { render, unmountComponentAtNode } from 'react-dom'
import EnterUsername from '../components/routes/EnterUsername'
import Message from '../components/chat/Message'
import ThemeContext from '../global/Theme'
import { ThemeProvider } from 'styled-components'
import FriendListItem, {
	WAITING_FOR_ACCEPT,
	WAITING_FOR_APPROVAL,
} from '../components/friends/FriendListItem'
import FriendResult from '../components/friends/FriendResult'
import { MESSAGES_GET, USER_EXISTS } from '../actions/types'
import { Reducer } from '../global/Reducer'


const theme = {
	color: {
		background: '#f7f7f7',
		windowBg: 'white',
		text: '#555',
		subText: '#777',
		primary: "#5CA4A9",
		primaryDark: "#44797d",
		secondary: "#0E7C7B",
		secondaryDark: '#0a5858',
		error: '#ED6A5A',
		accent: '#7E2E84',
		accentDark: '#542058',
		inputBorder: '#eeeeee',
		disabled: '#aaaaaa'
	},
	font: {
		title: {
			family: 'Roboto Slab',
			weight: 'bold'
		},
		body: {
			family: 'Open Sans',
			weight: 'normal',
		},
		size: {
			body: '1em',
			title: '1.5em',
			subTitle: '1em',
			pageHeader: '2em',
			button: '1.5em',
			input: '1.5em',
			small: '0.7em',
		}
	},
};

let state = {
	auth: {
		ID: null,
		username: null,
		password: null,
		exists: false
	},
	nav: {
		currentPage: '/enter-username',
		currentChat: null
	},
	friends: {
		incoming: [],
		outgoing: [],
		friends: []
	},
	messages: null
}

let container = null
beforeEach(() => {
	container = document.createElement('div')
	document.body.appendChild(container)
})

afterEach(() => {
	unmountComponentAtNode(container)
	container.remove()
	container = null
})

it('Renders correctly', () => {


	const message = renderer.create(<ThemeProvider theme={theme}><Message /></ThemeProvider>);
	expect(message.toJSON()).toMatchSnapshot();
});


it( "Displays correct friend list item status", () => {
	const friend = renderer.create(<ThemeProvider theme={theme}><FriendListItem status={WAITING_FOR_ACCEPT} /></ThemeProvider>);

	expect(friend.toJSON().children[1].children[0]).toBe('Accept');
} );


it( "Displays correct friend list item status", () => {
	const friend = renderer.create(<ThemeProvider theme={theme}><FriendListItem status={WAITING_FOR_APPROVAL} /></ThemeProvider>);

	expect(friend.toJSON().children[1].children[0]).toBe('Waiting...');
} );

it("Displays correct friend search result status", () => {
	const friend = renderer.create(<ThemeProvider theme={theme}><FriendResult status="loading" /></ThemeProvider>);

	expect(friend.toJSON().children[0].children[0].children[1].children[0].props.className).toBe("svg-inline--fa fa-spinner fa-w-16 fa-pulse ");
})


it("Displays correct friend search result status", () => {
	const friend = renderer.create(<ThemeProvider theme={theme}><FriendResult status="add" /></ThemeProvider>);

	expect(friend.toJSON().children[0].children[0].children[1].children[0]).toBe("Add Friend");
})

const stateAfterUserExists = () => {

	return Reducer(state, {
		type: USER_EXISTS,
		payload: {
			user_exists: true,
			username: 'test'
		}
	});
}

it("updates the username correctly after user exists returned", () => {
	state = stateAfterUserExists();

	expect(state.auth.username).toBe('test');
})



it("updates the current page correctly after user exists returned", () => {
	state = stateAfterUserExists();

	expect(state.nav.currentPage).toBe('/login');
})

it("updates the user_exists value correctly after user exists returned", () => {
	state = stateAfterUserExists();
	expect(state.auth.exists ).toBe(true);
})

