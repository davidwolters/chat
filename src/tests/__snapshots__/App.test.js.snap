// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div>
  <div
    className="sc-dIUggk bYBOI"
  >
    <div
      className="sc-idOhPF ioaswy"
    />
  </div>
</div>
`;
